rust/worker/src/execution/orchestration/metadata.rs

```diff
@@ -94,6 +117,321 @@ impl ChromaError for MetadataSegmentQueryError {
+    ///  wait for it to finish before returning the result.
+    pub(crate) async fn run(mut self) -> Result<usize, Box<dyn ChromaError>> {
+        let (tx, rx) = tokio::sync::oneshot::channel();
+        self.result_channel = Some(tx);
+        let mut handle = self.system.clone().start_component(self);
+        let result = rx.await;
+        handle.stop();
+        result.unwrap()
+    }
+}
+
+#[async_trait]
+impl Component for CountQueryOrchestrator {
+    fn get_name() -> &'static str {
+        "Count Query Orchestrator"
+    }
+
+    fn queue_size(&self) -> usize {
+        1000 // TODO: make this configurable
+    }
+
+    async fn on_start(&mut self, ctx: &crate::system::ComponentContext<Self>) -> () {
+        self.start(ctx).await;
+        self.pull_logs(ctx).await;
+    }
+}
+
+#[async_trait]
+impl Handler<PullLogsResult> for CountQueryOrchestrator {
+    async fn handle(&mut self, message: PullLogsResult, ctx: &ComponentContext<Self>) {
+        match message {
+            Ok(logs) => {
+                let logs = logs.logs();
+                println!("Got logs {}", logs.total_len());
+                println!("Logs {:?}", logs);
+                self.log_record_count = logs.total_len();
+                // TODO: Add logic for merging logs with count from record segment.
+                let operator = CountRecordsOperator::new();
+                let input = CountRecordsInput::new(
+                    self.record_segment
+                        .as_ref()
+                        .expect("Expect segment")
+                        .clone(),
+                    self.blockfile_provider.clone(),
+                );
+                let msg = wrap(operator, input, ctx.sender.as_receiver());
+                match self.dispatcher.send(msg, None).await {
+                    Ok(_) => (),
+                    Err(e) => {
+                        // Log an error - this implies the dispatcher was dropped somehow
+                        // and is likely fatal
+                        println!("Error sending Count Query task: {:?}", e);
+                    }
+                }
+            }
+            Err(e) => {
+                self.terminate_with_error(Box::new(e), ctx);
+            }
+        }
+    }
+}
+
+#[async_trait]
+impl Handler<Result<CountRecordsOutput, MergeMetadataResultsOperatorError>>
+    for CountQueryOrchestrator
+{
+    async fn handle(
+        &mut self,
+        message: Result<CountRecordsOutput, MergeMetadataResultsOperatorError>,
+        ctx: &ComponentContext<Self>,
+    ) {
+        let msg = match message {
+            Ok(m) => m,
+            Err(e) => {
+                return self.terminate_with_error(Box::new(e), ctx);
+            }
+        };
+        let channel = self
+            .result_channel
+            .take()
+            .expect("Expect channel to be present");
+        match channel.send(Ok(msg.count + self.log_record_count)) {
```

Comment:
We need to do some overlay here, right? The logrecords may be updates or deletes in which case we can't just add these two numbers if I understand correctly. Or are we planning to do that as follow-up (fine if so)?

Reply:
Yes follow-up. I mentioned in the PR description. Even for adds we can't just add because of what we discussed in the morning today if you remember